src/rust/lib/fuzzly/src/score.rs

```diff
@@ -0,0 +1,319 @@
+                Some(prev)                                 => Some(prev),
+                None                                       => Some((vertex,score)),
+            }
+        });
+        best_pair.map(|(vertex,_)| vertex)
+    }
+
+    fn best_path_rev(&self, end:subsequence_graph::Vertex) -> BestPathRevIter {
+        BestPathRevIter {
+            scores      : self,
+            next_vertex : Some(end),
+        }
+    }
+}
+
+struct BestPathRevIter<'a> {
+    scores      : &'a VerticesScores,
+    next_vertex : Option<subsequence_graph::Vertex>
+}
+
+impl<'a> Iterator for BestPathRevIter<'a> {
+    type Item = subsequence_graph::Vertex;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = std::mem::take(&mut self.next_vertex);
+        self.next_vertex = (|| {
+            let VerticesScores(scores) = self.scores;
+            Some(scores.get(&next?)?.best_input_path?.from)
+        })();
+        next
+    }
+}
+
+
+
+// ===================
+// === Score Match ===
+// ===================
+
+/// Fast-check if the pattern matches text.
+///
+/// This is faster way than calling `score_match(text,pattern,metric).is_some()`, therefore it's
+/// recommended to call this function before scoring when we are not sure if the pattern actually
+/// matches the text.
+pub fn matches(text:impl Str, pattern:impl Str) -> bool {
+    let mut pattern_chars     = pattern.as_ref().chars();
+    let mut next_pattern_char = pattern_chars.next();
+    for text_char in text.as_ref().chars() {
+        if let Some(pattern_char) = next_pattern_char {
+            if pattern_char.eq_ignore_ascii_case(&text_char) {
+                next_pattern_char = pattern_chars.next()
+            }
+        } else {
+            break;
+        }
+    }
+    next_pattern_char.is_none()
+}
+
+/// The result of `find_best_subsequence` function.
+#[derive(Clone,Debug,Default,PartialEq)]
+pub struct Subsequence {
+    /// The score of found subsequence.
+    pub score:f32,
+    /// Indices of `text`'s chars which belong to the subsequence.
+    pub indices:Vec<usize>
+}
+
+impl Subsequence {
+    /// Compare scores of subsequences.
+    ///
+    /// The `f32` does not implement total ordering, however that does not help when we want to
+    /// sort items by their matching score. Therefore this function assumes that all NaNs are the
+    /// lowest values.
+    pub fn compare_scores(&self, rhs:&Subsequence) -> std::cmp::Ordering {
+        if self.score.is_nan() && rhs.score.is_nan() {
+            std::cmp::Ordering::Equal
+        } else if self.score.is_nan() {
+            std::cmp::Ordering::Less
+        } else if rhs.score.is_nan() {
+            std::cmp::Ordering::Greater
+        } else if self.score < rhs.score {
+            std::cmp::Ordering::Less
+        } else if self.score > rhs.score {
+            std::cmp::Ordering::Greater
+        } else {
+            std::cmp::Ordering::Equal
+        }
```

Comment:
I'd write the bodies inline and align them - would be much more clear looking at the code imo